drgn_tools/crash_net.py

```diff
@@ -0,0 +1,372 @@
+            rows.append(
+                [
+                    neigh_addr,
+                    ip_addr,
+                    hw_type,
+                    hw_addr,
+                    dev_name,
+                    dev_state,
+                ]
+            )
+            neigh = neigh.next
+    print_table(rows)
+
+
+def print_task_sockets(prog: Program, pid: int, print_full_data: bool) -> None:
+    """
+    Displays the socket addreses, family type, source and destination
+    addresses and corresponding ports along with the dump of data
+    structures. If print_full_data is true, it prints detailed data
+    structure dumps.
+    :param prog: drgn program
+    :param pid: pid of the process.
+    :returns: None.
+    """
+
+    task = find_task(prog, int(pid))
+    if not (task):
+        print("Invalid PID.")
+        return
+
+    print(
+        "PID: "
+        + str(pid)
+        + "    TASK: "
+        + hex(task.value_())[2:]
+        + "    CPU: "
+        + str(int(task.cpu))
+        + "    COMMAND: "
+        + task.comm.string_().decode("utf-8")
+    )
+
+    if print_full_data:
+        print(
+            "%-5s %-20s %-20s %-20s %-20s %-20s"
+            % (
+                "FD",
+                "SOCKET",
+                "SOCK",
+                "FAMILY:TYPE",
+                "SOURCE-PORT",
+                "DESTINATION-PORT",
+            )
+        )
+
+    rows = [
+        [
+            "FD",
+            "SOCKET",
+            "SOCK",
+            "FAMILY:TYPE",
+            "SOURCE-PORT",
+            "DESTINATION-PORT",
+        ]
+    ]
+
+    for i in range(int(task.files.fdt.max_fds)):
+        try:
+            socket = SOCKET_I(task.files.fdt.fd[i].f_path.dentry.d_inode)
+            sock_family = str(
+                sockfamily_dict.get(int(socket.sk.__sk_common.skc_family))
+            )
+            if (
+                int(socket.sk.__sk_common.skc_family)
+            ) not in sockfamily_dict.keys():
+                sock_family = "AF_"
+            sock_family = sock_family[3:]
+            sock_type = str(socktype_dict.get(int(socket.type)))
+            if int(socket.type) not in socktype_dict.keys():
+                sock_type = "SOCK_"
+            sock_type = sock_type[5:]
+
+            if sock_family == "INET":
+                saddr = int(socket.sk.__sk_common.skc_rcv_saddr)
+                daddr = int(socket.sk.__sk_common.skc_daddr)
+                dport = socketlib.ntohs(int(socket.sk.__sk_common.skc_dport))
+                sport = int(socket.sk.__sk_common.skc_num)
+                source_port = (
+                    str(ipaddress.ip_address(saddr).reverse_pointer[:-13])
+                    + "-"
+                    + str(sport)
+                )
+                destination_port = (
+                    ipaddress.ip_address(daddr).reverse_pointer[:-13]
+                    + "-"
+                    + str(dport)
```

Comment:
Again, the IP address reading is broken here, due to the same endianness concerns described in the previous revisions. And further, the ntohs() is only necessary on little endian platforms.

If it helps, you could make a little helper that takes a pointer to an IPv4 address and returns the Python IPv4Address data structure for it:

```python
def get_ipv4(ptr: Object) -> ipaddress.IPv4Address:
    return ipaddress.IPv4Address(prog.read_bytes(ptr, 4))

skc = socket.sk.__sk_common
saddr = get_ipv4(skc.skc_rcv_saddr.address_of_())
daddr = get_ipv4(skc.skc_daddr.address_of_())
sport = int(skc.skc_num)
dport = int(sokcet.skc_dport)
if prog.platform.flags & PlatformFlags.IS_LITTLE_ENDIAN:
    dport = socketlib.ntohs(dport)

source_port = f"{saddr}:{sport}"
destination_port = f"{daddr}:{dport}"
```